test/protocol-tests/src/test/java/software/amazon/awssdk/protocol/tests/connection/SyncClientConnectionTest.java

```diff
@@ -0,0 +1,140 @@
+    private MockHttpClient mockHttpClient;
+
+    @Before
+    public void setupClient() {
+        mockHttpClient = new MockHttpClient();
+        client = ProtocolRestJsonClient.builder()
+                                       .credentialsProvider(StaticCredentialsProvider.create(AwsBasicCredentials.create("akid",
+                                                                                                                       "skid")))
+                                       .region(Region.US_EAST_1)
+                                       .httpClient(mockHttpClient)
+                                       .build();
+    }
+
+    @Test
+    public void nonStreaming_exception_shouldCloseConnection() throws IOException {
+        ClosableStringInputStream inputStream = new ClosableStringInputStream("{\"__type\":\"SomeUnknownType\"}");
+        mockHttpClient.stubNextResponse(mockResponse(inputStream, 400));
+
+        assertThatThrownBy(() -> client.allTypes(AllTypesRequest.builder().build()))
+            .isExactlyInstanceOf(ProtocolRestJsonException.class);
+        assertThat(inputStream.isClosed()).isTrue();
+    }
+
+    @Test
+    public void nonStreaming_successfulResponse_shouldCloseConnection() {
+        ClosableStringInputStream inputStream = new ClosableStringInputStream("{}");
+        mockHttpClient.stubNextResponse(mockResponse(inputStream, 200));
+        client.allTypes(AllTypesRequest.builder().build());
+        assertThat(inputStream.isClosed()).isTrue();
+    }
+
+    @Test
+    public void streamingOut_successfulResponse_shouldCloseConnection() {
+        ClosableStringInputStream inputStream = new ClosableStringInputStream("{}");
+        mockHttpClient.stubNextResponse(mockResponse(inputStream, 200));
+
+        client.streamingOutputOperation(b -> b.build(), ResponseTransformer.toBytes());
+        assertThat(inputStream.isClosed()).isTrue();
+    }
+
+    @Test
+    public void streamingOut_errorResponse_shouldCloseConnection() {
+        ClosableStringInputStream inputStream = new ClosableStringInputStream("{\"__type\":\"SomeUnknownType\"}");
+        mockHttpClient.stubNextResponse(mockResponse(inputStream, 400));
+
+        assertThatThrownBy(() -> client.streamingOutputOperation(b -> b.build(), ResponseTransformer.toBytes()))
+            .isExactlyInstanceOf(ProtocolRestJsonException.class);
+        assertThat(inputStream.isClosed()).isTrue();
+    }
+
+    @Test
+    public void streamingOut_connectionLeftOpen_shouldNotCloseStream() {
+        ClosableStringInputStream inputStream = new ClosableStringInputStream("{}");
+        mockHttpClient.stubNextResponse(mockResponse(inputStream, 200));
+
+        client.streamingOutputOperation(b -> b.build(), new ResponseTransformer<StreamingOutputOperationResponse, Object>() {
+
+            @Override
+            public Object transform(StreamingOutputOperationResponse response, AbortableInputStream inputStream) throws Exception {
```

Comment:
What do you think about adding a similar test where we return the `inputStream`, and then call `close()` and then check to see that the wrapped `InputStream` is closed?

We could look into a similar test for async

Reply:
+1, will add that test.

I thought about adding a similar test for async but unlike sync where connections are being closed by closing the inputStream, channels are being released in the netty module so there's no easy way to verify that here.